ovn-org/ovn-kubernetes
# Taint nodes with a too small MTU

**- What this PR does and why is it needed**
This PR taints nodes, if the nodes default interface MTU is not big enough to contain the Geneve header and the configured MTU (`config.Default.MTU`).

The taint has the key `k8s.ovn.org/mtu-too-small` and `NoSchedule` effect.

**- Special notes for reviewers**
I cherry-picked @tssurya s commit 28babd42 from #2183, so you might see some changes twice.


**- How to verify it**
For most parts I included unit tests. In addition an e2e test was included to test the whole procedure. 
To verify it locally, the MTU of one nodes default interface can be adjusted and the ovnkube-node pod restarted to trigger the validation (as done in the following with node ovn-worker2).

1. Get all pods in ovn-kubernetes namespace
```
$ k get po -o wide
NAME                              READY   STATUS    RESTARTS   AGE   IP           NODE                NOMINATED NODE   READINESS GATES
ovnkube-db-7ccf8cf86c-5f6r8       2/2     Running   0          11m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-master-75cfc8566c-hlrzm   3/3     Running   0          11m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-node-mm6b8                3/3     Running   1          11m   172.18.0.3   ovn-worker          <none>           <none>
ovnkube-node-pd6kj                3/3     Running   1          11m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-node-shz2q                3/3     Running   1          11m   172.18.0.2   ovn-worker2         <none>           <none>
ovs-node-d9wfz                    1/1     Running   0          11m   172.18.0.4   ovn-control-plane   <none>           <none>
ovs-node-kzs7k                    1/1     Running   0          11m   172.18.0.3   ovn-worker          <none>           <none>
ovs-node-sf49v                    1/1     Running   0          11m   172.18.0.2   ovn-worker2         <none>           <none>
```
2. Validate ovn-worker2 is not tainted:
```
$ k get no ovn-worker2 -o jsonpath='{.spec.taints[*]}'
```
3. Check the logs of the ovnkube-node pod running on ovn-worker2 about MTU actions:
```
$ k logs ovnkube-node-shz2q -c ovnkube-node| grep mtu
...
I0714 07:25:33.862117    1475 node.go:551] MTU (1500) is big enough to deal with Geneve header overhead (sum 1458). Making sure node is not tainted with {k8s.ovn.org/mtu-too-small  NoSchedule <nil>}...
```
4. Set the MTU of ovn-worker2s default interface very low: 
```
$ docker exec ovn-worker2 ip link show breth0 
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
$ docker exec ovn-worker2 ip link set breth0 mtu 1000 
$ docker exec ovn-worker2 ip link show breth0        
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1000 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
```
5. Restart the ovnkube-node pod from ovn-worker2
```
$ k delete po ovnkube-node-shz2q
pod "ovnkube-node-shz2q" deleted
$ k get po -o wide -w
NAME                              READY   STATUS    RESTARTS   AGE   IP           NODE                NOMINATED NODE   READINESS GATES
ovnkube-db-7ccf8cf86c-5f6r8       2/2     Running   0          13m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-master-75cfc8566c-hlrzm   3/3     Running   0          13m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-node-mm6b8                3/3     Running   1          13m   172.18.0.3   ovn-worker          <none>           <none>
ovnkube-node-pd6kj                3/3     Running   1          13m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-node-vggg9                1/3     Running   0          10s   172.18.0.2   ovn-worker2         <none>           <none>
ovs-node-d9wfz                    1/1     Running   0          13m   172.18.0.4   ovn-control-plane   <none>           <none>
ovs-node-kzs7k                    1/1     Running   0          13m   172.18.0.3   ovn-worker          <none>           <none>
ovs-node-sf49v                    1/1     Running   0          13m   172.18.0.2   ovn-worker2         <none>           <none>

ovnkube-node-vggg9                2/3     Running   0          32s   172.18.0.2   ovn-worker2         <none>           <none>
ovnkube-node-vggg9                3/3     Running   0          48s   172.18.0.2   ovn-worker2         <none>           <none>
^C%        
```
6. Check the logs from the restarted ovnkube-node pod form MTU actions:
```
$ k logs ovnkube-node-vggg9 -c ovnkube-node| grep mtu
...
I0714 07:38:44.773637    6259 node.go:547] MTU (1000) is not big enough to deal with Geneve header overhead (sum 1458). Tainting node with {k8s.ovn.org/mtu-too-small  NoSchedule <nil>}...
```
7. Validate that ovn-worker2 got tainted:
```
$ k get no ovn-worker2 -o jsonpath='{.spec.taints[*]}'
{"effect":"NoSchedule","key":"k8s.ovn.org/mtu-too-small"}
```
8. Reset the MTU of ovn-worker2 to its original value, restart the ovnkube-node pod and validate that the taint got removed from ovn-worker2 again
```
$ docker exec ovn-worker2 ip link show breth0        
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1000 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
$ docker exec ovn-worker2 ip link set breth0 mtu 1500
$ docker exec ovn-worker2 ip link show breth0        
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
$ k delete po ovnkube-node-vggg9
pod "ovnkube-node-vggg9" deleted
$ k get po -o wide -w                        
NAME                              READY   STATUS    RESTARTS   AGE   IP           NODE                NOMINATED NODE   READINESS GATES
ovnkube-db-7ccf8cf86c-5f6r8       2/2     Running   0          26m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-master-75cfc8566c-hlrzm   3/3     Running   0          26m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-node-mm6b8                3/3     Running   1          26m   172.18.0.3   ovn-worker          <none>           <none>
ovnkube-node-pd6kj                3/3     Running   1          26m   172.18.0.4   ovn-control-plane   <none>           <none>
ovnkube-node-tjstv                1/3     Running   0          11s   172.18.0.2   ovn-worker2         <none>           <none>
ovs-node-d9wfz                    1/1     Running   0          26m   172.18.0.4   ovn-control-plane   <none>           <none>
ovs-node-kzs7k                    1/1     Running   0          26m   172.18.0.3   ovn-worker          <none>           <none>
ovs-node-sf49v                    1/1     Running   0          26m   172.18.0.2   ovn-worker2         <none>           <none>

ovnkube-node-tjstv                2/3     Running   0          53s   172.18.0.2   ovn-worker2         <none>           <none>
ovnkube-node-tjstv                3/3     Running   0          78s   172.18.0.2   ovn-worker2         <none>           <none>
^C%
$ k logs ovnkube-node-tjstv -c ovnkube-node| grep mtu
...
I0714 07:51:24.751469   10599 node.go:551] MTU (1500) is big enough to deal with Geneve header overhead (sum 1458). Making sure node is not tainted with {k8s.ovn.org/mtu-too-small  NoSchedule <nil>}...
$ k get no ovn-worker2 -o jsonpath='{.spec.taints[*]}'
```

**- Description for the changelog**
Taint nodes with a too small MTU with `k8s.ovn.org/mtu-too-small`
